src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -21706,6 +21706,484 @@ void M(string? s)
+}";
+            var source = @"
+class D
+{
+    static void M(C c, string? p1, string? p2)
+    {
+        _ = p1 ?? throw null!;
+        c.M(p1, out var x1);
+        x1.ToString();
+
+        c.M(p2, out var x2);
+        x2.ToString(); // 1
+    }
+}";
+            var expected = new[]
+            {
+                // (11,9): warning CS8602: Dereference of a possibly null reference.
+                //         x2.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x2").WithLocation(11, 9)
+            };
+
+            var lib = CreateNullableCompilation(new[] { NotNullIfNotNulltAttributeDefinition, lib_cs });
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics(expected);
+
+            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, NotNullIfNotNulltAttributeDefinition });
+            comp2.VerifyDiagnostics(expected);
+            CompileAndVerify(comp2);
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_OutParameter_WithMaybeNullWhen()
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public bool M(string? p, [NotNullIfNotNull(""p""), MaybeNullWhen(true)] out string p2) => throw null!;
+}";
+            var source = @"
+class D
+{
+    static void M(C c, string? p1, string? p2)
+    {
+        _ = p1 ?? throw null!;
+        _ = c.M(p1, out var x1) 
+            ? x1.ToString()
+            : x1.ToString();
+
+        _ = c.M(p2, out var x2)
+        ? x2.ToString() // 1
+        : x2.ToString();
+    }
+}";
+            var expected = new[]
+            {
+                // (12,11): warning CS8602: Dereference of a possibly null reference.
+                //         ? x2.ToString() // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x2").WithLocation(12, 11)
+            };
+
+            var lib = CreateNullableCompilation(new[] { NotNullIfNotNulltAttributeDefinition, MaybeNullWhenAttributeDefinition, lib_cs });
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics(expected);
+
+            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, NotNullIfNotNulltAttributeDefinition, MaybeNullWhenAttributeDefinition });
+            comp2.VerifyDiagnostics(expected);
+            CompileAndVerify(comp2);
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_RefParameter()
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public void M(string? p, [NotNullIfNotNull(""p"")] ref string? p2) => throw null!;
+}";
+            var source = @"
+class D
+{
+    static void M(C c, string? p1, string? p2)
+    {
+        _ = p1 ?? throw null!;
+        string? x1 = null;
+        c.M(p1, ref x1);
+        x1.ToString();
+
+        string? x2 = null;
```

Comment:
Consider also testing where the input is non-null going in but null coming out.

Reply:
I didn't understand. Isn't that the scenario covered in the four preceding lines?

---
In reply to: [303566303](https://github.com/dotnet/roslyn/pull/37201#discussion_r303566303) [](ancestors = 303566303)